estimagic/logging/create_database.py

```diff
@@ -0,0 +1,253 @@
+
+Note: Most functions in this module change their arguments in place since this is the
+recommended way of doing things in sqlalchemy and makes sense for database code.
+
+"""
+from pathlib import Path
+
+from sqlalchemy import Column
+from sqlalchemy import create_engine
+from sqlalchemy import event
+from sqlalchemy import Float
+from sqlalchemy import Integer
+from sqlalchemy import MetaData
+from sqlalchemy import PickleType
+from sqlalchemy import String
+from sqlalchemy import Table
+from sqlalchemy.dialects.sqlite import DATETIME
+
+from estimagic.logging.update_database import append_rows
+
+
+def load_database(path, replace=False):
+    """Return database metadata object for the database stored in ``path``.
+
+    This is the default way of loading a database for read-only purposes in estimagic.
+
+    Args:
+        path (str or pathlib.Path): location of the database file. If the file does
+            not exist, it will be created.
+        replace (bool): If true and the database exists, it will be overwritten.
+            Otherwise, data will be appended.
+
+    Returns:
+        database (sqlalchemy.MetaData). The engine that connects to the database can be
+        accessed via ``database.bind``.
+
+    """
+    if isinstance(path, str):
+        path = Path(path)
+
+    if path.exists() and replace:
+        path.unlink()
+
+    engine = create_engine(f"sqlite:///{path}")
+    _make_engine_thread_safe(engine)
+    database = MetaData()
+    database.bind = engine
+    database.reflect()
+    return database
+
+
+def _make_engine_thread_safe(engine):
+    """Make the engine even more thread safe than by default.
+
+    The code is taken from the documentation: https://tinyurl.com/u9xea5z
+
+    The main purpose is to emit the begin statement of any connection
+    as late as possible in order to keep the time in which the database
+    is locked as short as possible.
+
+    """
+
+    @event.listens_for(engine, "connect")
+    def do_connect(dbapi_connection, connection_record):
+        # disable pysqlite's emitting of the BEGIN statement entirely.
+        # also stops it from emitting COMMIT before absolutely necessary.
+        dbapi_connection.isolation_level = None
+
+    @event.listens_for(engine, "begin")
+    def do_begin(conn):
+        # emit our own BEGIN
+        conn.execute("BEGIN DEFERRED")
+
+
+def prepare_database(
+    path,
+    params,
+    db_options=None,
+    optimization_status="scheduled",
+    gradient_status=0,
+    replace=False,
+):
+    """Return database metadata object with all relevant tables for the optimization.
+
+    This should always be used to create entirely new databases or to create the
+    tables needed during optimization in an existing database.
+
+    A new database is created if path does not exist yet or replace=True. Otherwise the
+    existing database is loaded and all tables needed to log the optimization are
+    overwritten.
```

Comment:
Here only the tables estimagic knows about will be re-created, but the database can have other tables. I clarified the docstring and removed the "replace" Argument. It is not really necessary, since we switched to